pyramid/config/__init__.py

```diff
@@ -903,6 +921,13 @@ def end(self):
         """
         return self.manager.pop()
 
+    def __enter__(self):
+        self.begin()
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        self.end()
```

Comment:
> What do you want to signify by clearing the action state?

Just that any work set up to be done within the scope of the `with` block is cancelled.

> Also much of the third party extensions I've ever seen do all their work in the directive without deferring it to the commit cycle.

https://youtu.be/QDvQ77JP8nw?t=50